### PR TITLE
SearchService changes for Search-by-TFM

### PR DESCRIPTION
**Pull request by advay26**
Addresses https://github.com/NuGet/Engineering/issues/4578

This change adds framework data to the Search Index through the Db2AzureSearch job:
* Accept new Frameworks and Tfms query parameters
* Construct filter strings with the selected Frameworks and Tfms
* Read new index fields in search response
* Add tests

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs [hunk 1]
```diff
@@ -10,5 +10,7 @@ public class V2SearchRequest : SearchRequest
         public V2SortBy SortBy { get; set; }
         public bool LuceneQuery { get; set; }
         public string PackageType { get; set; }
+        public string Frameworks { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs:13]: This model is essentially the service layer model, meaning it is best for it to most accurately represent the respective data types. In other words, I think Frameworks and Tfms should be collection data types here (e.g. List) not a string that gets parsed later.

Consider `V2SortBy`. This is a string in the controller but is immediately parsed to an enum in the controller (the "view" layer in an API). 

Could you try the same for `Frameworks` and `Tfms`?

The meta point is that it's easier to reason about the code if you can make the data types match their True representation as soon as possible.

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs:13]: I personally like `IReadOnlyList` for this purpose since we know it won't need to be modified after initialization.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 2]
```diff
@@ -147,7 +148,9 @@ private static void ApplyPaging(SearchOptions searchParameters, SearchRequest re
             SearchOptions searchParameters,
             SearchRequest request,
             bool isDefaultSearch,
-            string packageType)
+            string packageType,
+            string frameworks = null,
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:152]: In light of the comment above, this should become a list and probably be empty when the parameter is not included.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 3]
```diff
@@ -223,5 +236,18 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
 
             return orderBy;
         }
+
+        // Constructs filter strings for both Frameworks and Tfms.
+        // indexField: Determines which field you are targeting
+        //             i.e. IndexFields.Search.Frameworks or IndexFields.Search.Tfms
+        // frameworks: Comma-separated list of a user's selected Frameworks or Tfms
+        private string GetFrameworksOrTfmsFilterString(string indexField, string frameworks)
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:244]: This appears to have not sanitation or validation. I believe a user could craft a framework value that ends up being parsed as filter syntax not framework values (a la SQL injection).

How can we validate the frameworks and TFM values to ensure they are safe and only include expected characters?

**Review by joelverhagen — CHANGES_REQUESTED** [review]
I left some comments about input validation and data types for service layer models.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs (hunk 3 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs]: This makes sense. I've added the same validation steps we use when we're writing values into the index: checking the framework generations against the 4 supported constants, and checking that the TFMs are supported frameworks.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 4]
```diff
@@ -147,7 +149,9 @@ private static void ApplyPaging(SearchOptions searchParameters, SearchRequest re
             SearchOptions searchParameters,
             SearchRequest request,
             bool isDefaultSearch,
-            string packageType)
+            string packageType,
+            IReadOnlyList<string> frameworks = default,
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:153]: nit: IMO there is not much value in optional parameters on private methods. It's probably better to be explicit about parameters at all call sites.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 5]
```diff
@@ -223,5 +237,44 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
...
+            return filterStrings.Count() == 0
+                                ? String.Empty
+                                : " and (" + String.Join(" and ", filterStrings) + ")";
+        }
+
+        private bool IsValidFrameworkOrTfm(string indexField, string framework)
+        {
+            if (indexField == IndexFields.Search.Frameworks)
+            {
+                if (framework == AssetFrameworkHelper.FrameworkGenerationIdentifiers.Net ||
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:260]: consider a `static readonly` hash set which is a little easier to maintain.

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs [hunk 6]
```diff
@@ -85,6 +87,8 @@ public async Task<ActionResult<SearchStatusResponse>> GetStatusAsync()
                 SortBy = ParameterUtilities.ParseV2SortBy(sortBy),
                 LuceneQuery = luceneQuery ?? true,
                 PackageType = packageType,
+                Frameworks = frameworks == null ? null : frameworks.Split(','),
```
Comment (joelverhagen) [inline src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs:90]: what if there are spaces before or after the comma separator? Or multiple consecutive spaces?

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs [hunk 7]
```diff
@@ -68,6 +68,8 @@ public async Task<ActionResult<SearchStatusResponse>> GetStatusAsync()
             string sortBy = null,
             bool? luceneQuery = true,
             string packageType = null,
+            string frameworks = null,
```
Comment (joelverhagen) [inline src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs:71]: This is the protocol change. We should update the README either in this PR or in a subsequent one.

File: tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs [hunk 8]
```diff
@@ -217,6 +225,11 @@ public async Task UsesProvidedParameters()
                 Assert.True(lastRequest.LuceneQuery);
                 Assert.True(lastRequest.ShowDebug);
                 Assert.Equal("dotnettool", lastRequest.PackageType);
+                Assert.Equal(1, lastRequest.Frameworks.Count);
+                Assert.Contains("netcoreapp", lastRequest.Frameworks);
+                Assert.Equal(2, lastRequest.Tfms.Count);
+                Assert.Contains("net5.0", lastRequest.Tfms);
```
Comment (joelverhagen) [inline tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs:231]: I think you can do `Equal` on sequences handles the ordering, length, and value comparison automatically. Your current test does not assert on ordering (not a hard requirement for ordering but it's probably best to assert the exact behavior until we have a reason to be permissive on ordering).

File: tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs [hunk 9]
```diff
@@ -203,7 +209,9 @@ public async Task UsesProvidedParameters()
                     sortBy: "lastEdited",
                     luceneQuery: true,
                     debug: true,
-                    packageType: "dotnettool");
+                    packageType: "dotnettool",
+                    frameworks: "netcoreapp",
+                    tfms: "net5.0,netstandard2.1");
```
Comment (joelverhagen) [inline tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs:214]: Consider testing some more edge cases for these values. This is an open API endpoint. People can put weird stuff here.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 10]
```diff
@@ -223,5 +237,44 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
...
+            var filterStrings = frameworks
+                                    .Where(f => IsValidFrameworkOrTfm(indexField, f))
+                                    .Select(f => $"{indexField}/any(f: f eq '{f.ToLowerInvariant()}')");
+
+            return filterStrings.Count() == 0
+                                ? String.Empty
+                                : " and (" + String.Join(" and ", filterStrings) + ")";
+        }
+
+        private bool IsValidFrameworkOrTfm(string indexField, string framework)
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:256]: For the `Tfms` flow, we're using the original (user provided) value not the normalized version. Consider changing `IsValidFrameworkOrTfm` to `NormalizeOrNull` method that normalizes the input and returns null on invalid, Then you can just filter null after calling it.

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:256]: BTW, the behavior we have here is to _ignore_ bad input instead of erroring. We could return an error to the user if the input is not valid instead of ignoring it silently.

In general search service ignores bad input which ... maybe isn't a great dev experience? We could opt to be more explicit here.

Thoughts?

File: src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs [hunk 11]
```diff
@@ -10,5 +12,7 @@ public class V2SearchRequest : SearchRequest
         public V2SortBy SortBy { get; set; }
         public bool LuceneQuery { get; set; }
         public string PackageType { get; set; }
+        public IReadOnlyList<string> Frameworks { get; set; }
+        public IReadOnlyList<string> Tfms { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs:16]: Should `Tfms` be a `NuGetFramework` not a `string`? This would essentially force us to do the `NuGetFramework` and would mean (hopefully) we're only doing this parsing once per request, per TFM string rather than potentially multiple times for validation then normalization.

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs [hunk 12]
```diff
@@ -85,6 +87,8 @@ public async Task<ActionResult<SearchStatusResponse>> GetStatusAsync()
                 SortBy = ParameterUtilities.ParseV2SortBy(sortBy),
                 LuceneQuery = luceneQuery ?? true,
                 PackageType = packageType,
+                Frameworks = frameworks == null ? null : frameworks.Split(','),
+                Tfms = tfms == null ? null : tfms.Split(','),
```
Comment (joelverhagen) [inline src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs:91]: Thoughts on passing down `Array.Empty` instead of null? Nulls mean null checking so using the shared empty array instance provided by `Array.Empty` can lead to safer and cleaner code sometimes.

File: tests/NuGet.Services.AzureSearch.Tests/SearchService/SearchParametersBuilderFacts.cs [hunk 13]
```diff
@@ -570,6 +585,28 @@ public abstract class BaseFacts
...
+                    "(tfms/any(f: f eq 'net5.0'))" },
+                new object[] { new List<string> {"netframework"}, new List<string> {"foo"},
+                    "(frameworks/any(f: f eq 'netframework'))" },
+                new object[] { new List<string> {"foo"}, new List<string> {"net40-client"},
+                    "(tfms/any(f: f eq 'net40-client'))" },
+                new object[] { new List<string> {"net", "netstandard"},
+                    new List<string> {"netcoreapp3.1"},
+                    "(frameworks/any(f: f eq 'net') and frameworks/any(f: f eq 'netstandard'))" +
+                    " and (tfms/any(f: f eq 'netcoreapp3.1'))" },
+                new object[] { new List<string> {"netframework"},
```
[the remark this answers is not in the document]
Comment (advay26) [reply tests/NuGet.Services.AzureSearch.Tests/SearchService/SearchParametersBuilderFacts.cs:603]: Instead of doing it at the stage where we construct the parameters, I'm now validating and normalizing the framework and tfm inputs at the Controller layer. This means that once the framework and tfm values reach here, they should already be normalized, so I've added more test cases, but in SearchControllerFacts.cs instead of here.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs (hunk 5 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs]: Done. This validation is now in ParameterUtilities.cs, at the Controller layer.

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs (hunk 6 above)
Comment (advay26) [reply src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs]: What we need to do here is split by commas, then trim excess whitespace on either side of an entry, and then remove any empty entries. `Split` allows us to specify StringSplitOptions to trim whitespace and remove empty entries, but Trimming Entries is only available in .NET 5 and later. Since we would need to split, **then** trim, and **then** remove empty entries, we essentially need to do these steps separately (not using the Split overloads). I've done this now, but as separate explicit steps (in ParameterUtilities.cs)

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs (hunk 7 above)
Comment (advay26) [reply src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs:71]: I've updated the Readme as part of this change. Let me know how it reads

File: src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs (hunk 11 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/V2SearchRequest.cs:16]: For now, I decided to keep this as a List of strings. In the Controller layer (See ParameterUtilities.cs), we can parse the filter inputs as Frameworks, validate them, and then write the normalized strings to the resulting List, all at once. We would only be doing this cast once for each entry in a request, and it would be somewhat easier to work with strings in the Service layer (allows us to reuse some filter string code between Frameworks and Tfms more easily, for example)

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (advay26) [reply src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs]: I think it makes sense to do all the parsing, validating, and normalizing all in one place, and the Controller layer seems a better place for this. Regarding your comment on erroring on invalid inputs, doing these checks at a layer closer to the requester also seems like a better choice. What pros and cons were you thinking of?

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs (hunk 10 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs]: I've moved all the validating and normalizing code over to the Controller layer (ParameterUtilities.cs), and it looks a bit different now, so I'm not sure how the first comment applies. We are still ignoring invalid inputs rather than erroring with the current code though. We can flag invalid framework or tfm strings when we're validating them in the Controller. Do we have other examples of invalid parameter errors in Search? I'd like to see what the UI experience (and the code) for this might look like

File: src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGet.Services.SearchService.Core/Controllers/SearchController.cs]: Pros of validation in the service layer: any user of the service layer gets the same validations for free. 

Pros of validation in the controller layer: you will probably need to do view-specific validation here anyway, it keeps your service layer clean of view-specific validations (such as comma stuff)

It's a little bit moot since we only have one user of the service layer (the controller)

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs [hunk 14]
```diff
@@ -24,6 +27,11 @@ public static class ParameterUtilities
             { "totalDownloads-desc", V2SortBy.TotalDownloadsDesc },
         };
 
+        private static readonly HashSet<string> FrameworkGenerationIdentifiers = new HashSet<string> { AssetFrameworkHelper.FrameworkGenerationIdentifiers.Net,
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:30]: consider using a case insensitive comparer for the HashSet ctor. Might be a nice little UX improvement for the API since NuGetFramework in general is case insensitive.
https://nugettools.azurewebsites.net/6.0.0/parse-framework?framework=NET46

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs [hunk 15]
```diff
@@ -45,5 +53,27 @@ public static bool ParseIncludeSemVer2(string semVerLevel)
...
+                                                                .ToList();
+        }
+
+        public static IReadOnlyList<string> ParseTfms(string tfms)
+        {
+            return tfms == null ? new List<string>() : tfms
+                                                        .Split(',')
+                                                        .Select(f => f.ToLowerInvariant().Trim())
+                                                        .Where(f => f != String.Empty)
+                                                        .Select(f => new PackageFramework() { TargetFramework = f })
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:73]: nit: You don't need `PackageFramework` at all right? Can't you use `NuGetFramework` directly?

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs [hunk 16]
```diff
@@ -45,5 +53,27 @@ public static bool ParseIncludeSemVer2(string semVerLevel)
                 return semVerLevelVersion >= SemVer2Level;
             }
         }
+
+        public static IReadOnlyList<string> ParseFrameworks(string frameworks)
+        {
+            return frameworks == null ? new List<string>() : frameworks
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:59]: You can use Array.Empty instead of the first option in the ternary. This will reduce allocations over creating a new empty list each time.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 17]
```diff
@@ -223,5 +236,19 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
 
             return orderBy;
         }
+
+        // Constructs filter strings for both Frameworks and Tfms.
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:240]: nit: put this in an XML doc block not a normal comment block

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs [hunk 18]
```diff
@@ -223,5 +236,19 @@ private static IList<string> GetOrderBy(V2SortBy sortBy)
...
+        // indexField: Determines which field you are targeting
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs:241]: there is a `<param>` XML doc thing for this.

File: src/NuGet.Services.SearchService.Core/README.md [hunk 19]
```diff
@@ -100,7 +100,7 @@ package IDs and package versions.
 This endpoint is used by NuGetGallery in several scenarios when the gallery is configured with an external search
 service. Therefore, the contract should be considered unstable by external clients and may change freely to match the
 requirements of NuGetGallery. We strongly urge external applications (even NuGet client) to avoid using this endpoint
-since it should be considered and implementation detail of NuGetGallery. External clients should use the document V3
+since it should be considered an implementation detail of NuGetGallery. External clients should use the document V3
```
Comment (joelverhagen) [inline src/NuGet.Services.SearchService.Core/README.md:103]: 👀 oops. thanks for this

File: tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs [hunk 20]
```diff
@@ -145,6 +145,8 @@ public async Task HasDefaultParameters()
                 Assert.True(lastRequest.LuceneQuery);
                 Assert.False(lastRequest.ShowDebug);
                 Assert.Null(lastRequest.PackageType);
+                Assert.Equal(new List<string>(), lastRequest.Frameworks);
```
Comment (joelverhagen) [inline tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs:148]: nit: consider Assert.Same on Array.Empty, this is a strong assertion and states we don't allocate a new list for empty frameworks.

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs [hunk 21]
```diff
@@ -45,5 +53,27 @@ public static bool ParseIncludeSemVer2(string semVerLevel)
...
+        }
+
+        public static IReadOnlyList<string> ParseTfms(string tfms)
+        {
+            return tfms == null ? new List<string>() : tfms
+                                                        .Split(',')
+                                                        .Select(f => f.ToLowerInvariant().Trim())
+                                                        .Where(f => f != String.Empty)
+                                                        .Select(f => new PackageFramework() { TargetFramework = f })
+                                                        .Where(f => f.FrameworkName.IsSpecificFramework && !f.FrameworkName.IsPCL)
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:74]: 🚫🥒
(PCLs are jokingly called pickles sometimes)

**Review by joelverhagen — APPROVED** [review]
Address comments then :shipit:

**Review by zhhyu — APPROVED** [review]

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 16 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:59]: List<string> does not cast easily to an Array. Should we be using Array<string> for Frameworks and Tfms in V2SearchRequest, and then move the code to use Arrays everywhere instead? https://github.com/NuGet/NuGet.Jobs/pull/1103#discussion_r998348368

Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:59]: Both are IReadOnlyList though. You can cast one of the ternary sides to `(IReadOnlyList<string>)` and it should work

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 14 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:30]: Our current implementation is case insensitive right now (we use `ToLower()` on the inputs, and compare them to the HashSet values, which are all lowercase).

Even if we did use a case insensitive comparer for the HashSet, we would still need to normalize the input string (using `ToLower`) after comparing it against the HashSet right? Functionally, I don't think there's any difference, but making the HashSet use a non-default comparer might actually slow down perf in this case

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 16 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:59]: Is casting an empty Array to a List this way better than allocating a new List directly?

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 15 above)
Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:73]: Yeah, I was struggling to parse the Tfm strings as NuGetFrameworks directly, so I was going through PackageFramework to do it. I've figured it out now, so I'll make that change in the next commit.

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 16 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:59]: Cast it to IReadOnlyList, not List.
https://dotnetfiddle.net/6nv6Za

And yes, it's better. Casting does not (typically) allocate on the heap.

File: src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs (hunk 10 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/SearchParametersBuilder.cs]: No examples right now. In general we don't error out on bad input, we ignore it. And the error would probably not be surfaced in the Gallery UI at all. It would be more for custom code against our APIs (3rd party devs)

File: tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs (hunk 20 above)
Comment (advay26) [reply tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs]: I realized there's also an `Assert.Empty`, which seems like the best option here. Does that work?

Comment (joelverhagen) [reply tests/NuGet.Services.SearchService.Core.Tests/Controllers/SearchControllerFacts.cs]: It works, however it doesn't assert the shared instance is used like `Assert.Same` does. Not a big deal though, I'll leave it up to you.

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs [hunk 22]
```diff
@@ -45,5 +55,31 @@ public static bool ParseIncludeSemVer2(string semVerLevel)
...
+
+        public static IReadOnlyList<string> ParseTfms(string tfms)
+        {
+            return tfms == null
+                            ? (IReadOnlyList<string>)Array.Empty<string>()
+                            : tfms
+                                .Split(',')
+                                .Select(f => f.Trim())
+                                .Where(f => f != String.Empty)
+                                .Select(f => NuGetFramework.Parse(f))
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs:79]: Take a look at the code of `NuGetFramework`. There are some cases that throw instead of returning unspecified.
https://github.com/NuGet/NuGet.Client/blob/d1620d3bc77030d05e7e76b5adf06c33c4dc8374/src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs#L137-L144

https://github.com/NuGet/NuGet.Client/blob/d1620d3bc77030d05e7e76b5adf06c33c4dc8374/src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs#L247-L250

We may need to catch `ArgumentException` 😨. UTs should probably cover these to verify behavior over time.

Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs]: I've done this and added tests as well. Let me know how it looks

**Review by joelverhagen — APPROVED** [review]

File: src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs (hunk 14 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs]: Great point! You could do an all in one with a case insensitive Dictionary instead of a dictionary where the keys are allowed inputs and the values are the normalized inputs. This means you're not allocating for the `ToLower` but we're talking about micro-optimizations here without any benchmark :) go for what is easiest to read.

Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs]: I ran the benchmarking tool on 
(1) the original method with ToLower and a HashSet, 
(2) a case insensitive Dictionary (with normalized strings as values) with no ToLower, and
(3) a regular (case sensitive) Dictionary with no ToLower (This would not normalize inputs correctly, but I wanted to check its perf. This should be and was the quickest of the 3 methods).

The original method (1) was quicker than the Dictionary method (2) by about 5%. I'm guessing that the case insensitive comparer slows down the Dictionary access by longer than the ToLower allocation does.
image

Comment (advay26) [reply src/NuGet.Services.AzureSearch/SearchService/Models/ParameterUtilities.cs]: I also added duplicate checks to the code now, thanks for calling it out.
